### PR TITLE
IDE: refactor USR generation to request evaluator model.

### PR DESCRIPTION
**Pull request by nkcsgexi**
This patch adds a new TypeID zone for IDE usage, which contains
request kinds to answer IDE-interested questions. As a proof-of-concept,
we refactor the existing USR generation code to use the request evaluator
model. Most part of this patch is NFC, however a slight difference is
using this model allows us to cache the calculated USRs in the evaluator.

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by jrose-apple** [comment]
Where does the name "TypeID" come from?

File: lib/AST/USRGeneration.cpp [hunk 1]
```diff
@@ -215,30 +221,41 @@ bool ide::printDeclUSR(const ValueDecl *D, raw_ostream &OS) {
...
+    return std::string();
 
   Mangle::ASTMangler NewMangler;
-  std::string Mangled = NewMangler.mangleDeclAsUSR(D, getUSRSpacePrefix());
-
-  OS << Mangled;
+  return NewMangler.mangleDeclAsUSR(D, getUSRSpacePrefix());
+}
 
+bool ide::printDeclUSR(const ValueDecl *D, raw_ostream &OS) {
```
Comment (DougGregor) [inline lib/AST/USRGeneration.cpp]: It seems like this belongs in the IDE library.

**Review by DougGregor — CHANGES_REQUESTED** [review]
I don't think there's any reason that this code (the IDE requests or their implementation) have to live in the AST library. Can you put all of it in the IDE library instead?

**Comment by nkcsgexi** [comment]
Talked with @DougGregor offline about this PR. Since the API `printDeclUSR` is used by `Serialization/ModuleFile.cpp` too, moving them entirely to IDE seems to be difficult. Instead, we drop the IDE requests ID zone for now and move the USR generation request to type checker requests.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
The linux failure seems to be unrelated. @swift-ci Please clean test Linux platform

**Comment by nkcsgexi** [comment]
@swift-ci Please clean test Linux platform

**Comment by nkcsgexi** [comment]
all tests passed. @DougGregor OK to merge?

**Pull request merged**